tests/qam-kgraft/update_kgraft.pm

```diff
@@ -0,0 +1,164 @@
+    script_run(qq{if \$(zypper lr | grep test-kgraft -q);then zypper rr test-kgraft ; fi });
+    #check kgr status
+    until (kgr_status) {
+        kgr_block;
+    }
+
+    # create reference snapshot
+    my $snapshot_before = "snap_before-" . timemark;
+    my $ret             = $svirt->run_cmd("virsh snapshot-create-as $name $snapshot_before");
+    die "snapshot $snapshot_before failed" if $ret;
+    set_var('KGRAFT_SNAPSHOT_BEFORE', $snapshot_before);
+
+    # check if automounter works
+    check_automounter;
+
+    # RUN HEAVY LOAD script
+    # TODO: place script somewhere in git and deploy from here
+    #       ie /var/tmp/scripty.sh ; applies for all used scripts
+    script_run("/root/bin/heavy_load.sh");
+    #
+    sleep 15;
+
+
+    #INSTALL UPDATE ... #TODO it needs some coop with IBS ( update repo , patch name)
+    my $repo = get_var('KGRAFT_TEST_REPO');
+    zypper_call("ar -f $repo test-kgraft");
+    zypper_call("ref");
+
+    # TODO . it needs patchinfo and definition of patch
+    #save patch name to VAR
+    script_run("zypper patches | awk '/test-kgraft/ { print \$3; }' | tee > /dev/$serialdev", 0);
+
+    my $out = wait_serial(qr/SUSE-*/);
+    diag($out);
+    set_var('KGRAFT_PATCH_NAME', $out);
+
+    #patch system
+    zypper_call(qq{in -l -y -t patch \$(zypper patches | awk -F "|" '/test-kgraft/ { print \$2;}')}, [0, 102, 103]);
+
+    zypper_call("rr test-kgraft");
+    # check if kgraft patch is applied to all functions..
+    until (kgr_status) {
+        kgr_block;
+    }
+
+    # again check automounter
+    check_automounter;
+
+    #kill HEAVY-LOAD scripts
+    script_run("screen -S LTP_syscalls -X quit");
+    script_run("screen -S newburn_KCOMPILE -X quit");
+    script_run("/root/bin/heavy_load--tidyup.sh");
+
+    # wait for cooldown:)
+    sleep 45;
+
+    # create snapshot after update
+    my $snapshot_after = "snap_after-" . timemark;
+    $ret = $svirt->run_cmd("virsh snapshot-create-as $name $snapshot_after");
+    die "snapshot $snapshot_after failed" if $ret;
+    set_var('KGRAFT_SNAPSHOT_AFTER', $snapshot_after);
+    capture_state("after");
+    type_string("logout\n");
+
+}
+
+sub post_fail_hook() {
+    my $self            = shift;
+    my $name            = get_var('VIRSH_GUESTNAME');
+    my $snapshot_before = get_var('KGRAFT_SNAPSHOT_BEFORE');
+    save_screenshot;
+    send_key('ctrl-c');
+    sleep 2;
+    script_run("screen -S LTP_syscalls -X quit");
+    script_run("screen -S newburn_KCOMPILE -X quit");
+    script_run("/root/bin/heavy_load--tidyup.sh");
+    capture_state("fail");
+    type_string("logout\n");
+
+    # reconnect to svirt console
+    my $svirt = select_console('svirt');
+    $svirt->attach_to_running($name);
+
+    # revert to snapshot before update if it exists
+    if ($snapshot_before) {
+        my $ret = $svirt->run_cmd("virsh snapshot-revert $name $snapshot_before --running");
+        die "Snapshot $snapshot_before failed" if $ret;
+    }
```

Comment:
[DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself). You are calling the same code multiple times, please [extract method](http://refactoring.com/catalog/extractMethod.html).